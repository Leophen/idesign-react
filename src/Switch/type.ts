export interface SwitchProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 开关尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 开关关闭时的颜色
   */
  inactiveColor?: string;
  /**
   * 开关打开时的颜色
   */
  activeColor?: string;
  /**
   * 开关关闭时的标签
   */
  inactiveLabel?: React.ReactNode;
  /**
   * 开关打开时的标签
   */
  activeLabel?: React.ReactNode;
  /**
   * 是否禁用组件
   * @default false
   */
  disabled?: boolean;
  /**
   * 是否为加载状态
   * @default false
   */
  loading?: boolean;
  /**
   * 开关固定值（受控）
   */
  value?: boolean;
  /**
   * 开关默认值（非受控）
   * @default false
   */
  defaultValue?: boolean;
  /**
   * 切换开关时触发
   */
  onChange?: (value: boolean) => void;
}
