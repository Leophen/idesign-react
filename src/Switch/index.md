---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 1
---

# Switch 开关

用于两个互斥选项，多用于触发「开/关」

## 基本用法

不带描述，最基础的开关。

```tsx
import React, { useState } from 'react'
import { Switch } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(true)

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <Switch />
      <h4>有默认值</h4>
      <Switch defaultValue={value} />
      <h4>有受控值</h4>
      <Switch value={value} />
      <h4>一般用法</h4>
      <Switch
        value={value}
        onChange={handleChange}
      />
    </>
  )
}

export default App
```

## 自定义颜色

提供 `inactiveColor` 和 `activeColor` 属性控制开关关闭和打开时的颜色。

```tsx
import React, { useState } from 'react';
import { Switch } from 'idesign-react';

const App = () => {
  const [checked, setChecked] = useState(true);

  const handleChange = (val) => {
    setChecked(val);
  };

  return (
    <div className="idesign-demo-block-row">
      <Switch inactiveColor="#65b687" value={checked} onChange={handleChange} />
      <Switch activeColor="#65b687" value={checked} onChange={handleChange} />
      <Switch
        inactiveColor="#ec7491"
        activeColor="#65b687"
        value={checked}
        onChange={handleChange}
      />
    </div>
  );
};

export default App;
```

## 不同尺寸

提供 `small`、`medium`（默认）、`large` 三种尺寸的开关。

```tsx
import React, { useState } from 'react';
import { Switch } from 'idesign-react';

const App = () => {
  const [checked, setChecked] = useState(true);

  const handleChange = (val) => {
    setChecked(val);
  };

  return (
    <div className="idesign-demo-block-row">
      <Switch size="small" value={checked} onChange={handleChange} />
      <Switch size="medium" value={checked} onChange={handleChange} />
      <Switch size="large" value={checked} onChange={handleChange} />
    </div>
  );
};

export default App;
```

## 带描述内容的开关

提供 `inactiveColor` 和 `activeColor` 属性控制开关关闭和打开时的颜色。

```tsx
import React, { useState } from 'react';
import { Switch } from 'idesign-react';
import { Icon } from 'idesign-react';

const App = () => {
  const [checked, setChecked] = useState(false);

  const handleChange = (val) => {
    setChecked(val);
  };

  const inactiveIcon = <Icon name="Close" />;
  const activeIcon = <Icon name="Check" />;

  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Switch
          size="small"
          inactiveLabel="关"
          activeLabel="开"
          value={checked}
          onChange={handleChange}
        />
        <Switch
          size="medium"
          inactiveLabel="关"
          activeLabel="开"
          value={checked}
          onChange={handleChange}
        />
        <Switch
          size="large"
          inactiveLabel="关"
          activeLabel="开"
          value={checked}
          onChange={handleChange}
        />
      </div>
      <div className="idesign-demo-block-row">
        <Switch
          size="small"
          inactiveLabel={inactiveIcon}
          activeLabel={activeIcon}
          value={checked}
          onChange={handleChange}
        />
        <Switch
          size="medium"
          inactiveLabel={inactiveIcon}
          activeLabel={activeIcon}
          value={checked}
          onChange={handleChange}
        />
        <Switch
          size="large"
          inactiveLabel={inactiveIcon}
          activeLabel={activeIcon}
          value={checked}
          onChange={handleChange}
        />
      </div>
    </div>
  );
};

export default App;
```

## 禁用状态

可以使用 `disabled` 属性来定义开关是否被禁用，该属性接受一个 `Boolean` 类型的值。

```tsx
import React, { useState } from 'react';
import { Switch } from 'idesign-react';

const App = () => {
  const [checked, setChecked] = useState(true);

  const onChange = (val) => {
    setChecked(val);
  };

  return (
    <div className="idesign-demo-block-row">
      <Switch disabled />
      <Switch disabled value={checked} onChange={onChange} />
    </div>
  );
};

export default App;
```

## 加载中的状态

提供 `loading` 属性控制开关是否为加载状态。

```tsx
import React, { useState } from 'react';
import { Switch } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Switch size="small" loading />
      <Switch size="medium" loading />
      <Switch size="large" loading value={true} />
    </div>
  );
};

export default App;
```

<API />
