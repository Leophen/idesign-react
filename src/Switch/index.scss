.i-switch {
  position: relative;
  display: inline-flex;
  align-items: center;
  min-width: 40px;
  height: 20px;
  margin: 0;
  padding: 0;
  line-height: 20px;
  vertical-align: middle;
  background-color: var(--i-info-active);
  border: 0;
  border-radius: 10px;
  outline: none;
  cursor: pointer;
  transition: all 0.2s cubic-bezier(0, 0, 0.15, 1);
  user-select: none;
  &:hover {
    .i-switch__handle {
      &::before {
        box-shadow: 0 0 3px 1px rgb(0 0 0 / 20%);
      }
    }
  }
  &:active:not(.i-switch-is-disabled) {
    .i-switch__handle::before {
      right: -8px;
      left: 0;
    }
  }
  &.i-switch-is-checked {
    background-color: var(--i-primary);
    .i-switch__handle {
      left: calc(100% - 2px);
      transform: translate(-100%);
      .icon-Loading {
        color: var(--i-primary-disabled) !important;
      }
    }
    &:active:not(.i-switch-is-disabled) {
      .i-switch__handle::before {
        right: 0;
        left: -8px;
      }
    }
  }
  &.i-switch-is-disabled {
    background-color: var(--i-info-hover);
    cursor: not-allowed;
    &.i-switch-is-checked {
      background-color: var(--i-primary-disabled);
    }
    .i-switch__handle {
      cursor: not-allowed;
    }
  }
}

@keyframes rotating {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}

.i-switch__handle {
  position: absolute;
  top: 2px;
  left: 2px;
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border-radius: 12px;
  transform: translate(0%);
  visibility: visible;
  cursor: pointer;
  transition: all 0.2s cubic-bezier(0.38, 0, 0.24, 1);
  user-select: none;
  .icon-Loading {
    z-index: 1;
    color: var(--i-info) !important;
    transform: scale(0.8);
    animation: rotating 2s linear infinite;
  }
  &::before {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: #fff;
    border-radius: 12px;
    box-shadow: 0 0 3px 1px rgb(0 0 0 / 10%);
    transition: all 0.2s cubic-bezier(0.38, 0, 0.24, 1);
    content: '';
  }
}

.i-switch__content {
  display: inline-flex;
  align-items: center;
  box-sizing: border-box;
  min-width: 36px;
  padding-left: 22px;
  color: #fff;
  font-size: 12px;
  transition: padding 0.2s cubic-bezier(0.38, 0, 0.24, 1), opacity 0.2s linear;
  .i-icon {
    margin-left: -1px;
    color: #fff !important;
    font-size: 14px !important;
  }
}

.i-switch {
  &.i-switch-is-checked {
    .i-switch__content {
      padding-left: 6px;
    }
  }
}

@mixin injectSize($size, $size2) {
  min-width: $size;
  height: calc($size / 2);
  border-radius: calc($size / 4);
  .i-switch__handle {
    width: $size2;
    height: $size2;
  }
  &:active:not(.i-switch-is-disabled) {
    .i-switch__handle::before {
      right: calc($size * -0.2);
    }
  }
  &.i-switch-is-checked {
    &:active:not(.i-switch-is-disabled) {
      .i-switch__handle::before {
        left: calc($size * -0.2);
      }
    }
  }
}

.i-switch--size-small {
  @include injectSize(32px, 12px);
  .i-switch__content {
    padding-left: 20px;
    transform: scale(0.9);
  }
  &.i-switch-is-checked {
    .i-switch__content {
      padding-left: 4px;
    }
  }
}

.i-switch--size-large {
  @include injectSize(48px, 20px);
  .i-switch__content {
    padding-left: 26px;
    transform: scale(1.1);
  }
  &.i-switch-is-checked {
    .i-switch__content {
      padding-left: 8px;
    }
  }
}
