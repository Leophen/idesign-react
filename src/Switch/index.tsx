import React from 'react';
import classNames from 'classnames';
import './index.scss';
import useDefault from '../hooks/useDefault';
import Icon from '../Icon';
import { SwitchProps } from './type';

const Switch: React.FC<SwitchProps> = (props) => {
  const {
    className,
    style,
    size = 'medium',
    disabled = false,
    loading = false,
    inactiveColor,
    activeColor,
    inactiveLabel,
    activeLabel,
    value,
    defaultValue = false,
    onChange = () => { },
    ...restProps
  } = props;

  const [innerChecked, setInnerChecked] = useDefault(value, defaultValue, onChange);

  const handleSwitch = () => {
    if (disabled || loading) return;
    const changedValue = !innerChecked;
    setInnerChecked(changedValue)
  };

  const renderContent = (value: string | number | boolean) => {
    if (!!value && activeLabel) {
      return activeLabel;
    }
    if (!value && inactiveLabel) {
      return inactiveLabel;
    }
    return null;
  };

  return (
    <button
      type="button"
      role="switch"
      disabled={disabled}
      className={classNames(
        'i-switch',
        innerChecked && 'i-switch-is-checked',
        (disabled || loading) && 'i-switch-is-disabled',
        size === 'small' && 'i-switch--size-small',
        size === 'large' && 'i-switch--size-large',
        className,
      )}
      style={{
        ...(style || {}),
        ...({
          backgroundColor: innerChecked ? activeColor : inactiveColor,
        } || {}),
      }}
      onClick={handleSwitch}
      {...restProps}
    >
      <span className={'i-switch__handle'}>
        {loading && <Icon name="Loading" size={{ small: 12, medium: 16, large: 20 }[size]} />}
      </span>
      {renderContent(innerChecked) && (
        <div className="i-switch__content">{renderContent(innerChecked)}</div>
      )}
    </button>
  );
};

Switch.displayName = 'Switch';

export default Switch;
