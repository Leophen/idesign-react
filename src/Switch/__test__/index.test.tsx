import { render, fireEvent } from '@testing-library/react';
import Switch from '../index';

describe('Switch 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Switch />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('value', () => {
    const { container } = render(<Switch value={true} />);
    expect(container.firstChild).toHaveClass('i-switch-is-checked');
  });

  it('defaultValue', () => {
    const { container } = render(<Switch defaultValue={true} />);
    expect(container.firstChild).toHaveClass('i-switch-is-checked');
  });

  it('onChange', () => {
    const changeFn = jest.fn();
    const { container } = render(<Switch onChange={changeFn} />);
    fireEvent.click(container.firstChild);
    expect(changeFn).toHaveBeenCalled();
  });

  it('inactiveColor', () => {
    const { container } = render(<Switch inactiveColor="blue" />);
    expect(container.firstChild).toHaveStyle('background-color: blue');
  });

  it('activeColor', () => {
    const { container } = render(<Switch activeColor="red" value={true} />);
    expect(container.firstChild).toHaveStyle('background-color: red');
  });

  it('size', () => {
    const { container } = render(<Switch size="small" />);
    expect(container.firstChild).toHaveClass('i-switch--size-small');
  });

  it('inactiveLabel', () => {
    const { getByText } = render(<Switch inactiveLabel="关" />);
    expect(getByText('关')).toBeInTheDocument();
  });

  it('activeLabel', () => {
    const { getByText } = render(<Switch activeLabel="开" value={true} />);
    expect(getByText('开')).toBeInTheDocument();
  });

  it('disabled', () => {
    const { container } = render(<Switch disabled />);
    expect(container.firstChild).toHaveClass('i-switch-is-disabled');
  });
});

