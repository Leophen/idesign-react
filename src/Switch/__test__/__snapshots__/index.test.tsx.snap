// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch 组件测试 create 1`] = `
<DocumentFragment>
  <button
    class="i-switch"
    role="switch"
    type="button"
  >
    <span
      class="i-switch__handle"
    />
  </button>
</DocumentFragment>
`;
